[1.3] odML tables integration

mpsonntag opened the pull request:
This PR introduces plugin support of [python-odmltables](https://github.com/INM-6/python-odmltables).
- It requires odmltables version 1.1.4, which depends on INM-6/python-odmltables#13. 
- If the required version could not be found, an appropriate informative message is displayed.
- The currently supported odmltables plugins are `compare`, `convert`, `filter` and `merge`.
- Currently odmltables only supports xml files with the file ending `.odml`. Therefore, any odML document that is passed from odml-ui to one of the supported odmltables wizards, is saved as an xml file with the ".odml" file ending in the temporary odML folder also used for the terminologies and then passed along to odmltables.

Depending on how the package is installed (`python setup.py install` vs `pip install`) and which environment is used (`virtualenv` vs `conda`)
- the license file and custom icons are put to different directories
- the gtk IconTheme might not be out of the box available

Therefore this PR
- Adds an additional `conda` dependency which adds all required gtk icons.
- Adds handling paths differently if `conda` or a `virtualenv` is used.
- Changes the odmltables icons from svg to png, since the `conda` gtk environment cannot handle svgs.
- Updates `setup.py` to autogenerate the `odmlui` startup script. With the previous version it happend, that `conda` environments and `virtual` environments, when both present, were crosspolluting the shebang line of the `odml-gui` startup script rendering them unusable.

Installation has been tested for any combination of `Python2` and  `Python3` with `virtualenv` and `Anaconda` and installation option `python setup.py install` and `pip install .`.

mpsonntag committed:
[Editor] add odml_tables attribute

mpsonntag committed:
[Editor] add odML tables compare

mpsonntag committed:
[Editor] add odML tables convert

mpsonntag committed:
[Editor] add odML tables filter

mpsonntag committed:
[Editor] add toolbar separators

mpsonntag committed:
[info] Add required odmltables version

mpsonntag committed:
[Editor] Update odmltables availability check

mpsonntag committed:
[Editor] Temp .odml files for odmltables

mpsonntag committed:
[Editor] Pep8 fixes of odMLTable entries

mpsonntag committed:
[Editor] Add odML Tables merge

mpsonntag committed:
[Editor] Remove odmltables py2 restriction

mpsonntag committed:
[Editor] Update odmltables requirement message

JuliaSprenger commented:
This version can not be run yet when using a conda environment and pip for installation. I get the following error when running the odml-gui: 

> .../odmlui/Editor.py", line 14, in <module>
>     from odml.terminology import CACHE_DIR
> ImportError: cannot import name 'CACHE_DIR'

mpsonntag commented:
Good catch, its missing requirement python-odml 1.3.3 to include the CACHE_DIR...

mpsonntag committed:
[setup] Update python-odml requirement

mpsonntag committed:
[odml-gui] Update python interpreter

Making sure that dependent on the environment
(virtualenv, conda) the proper python interpreter
is used.

mpsonntag committed:
[images] Add INM6 png images

Some Pixbuf versions do not support svg,
removed svgs and added pngs.

mpsonntag committed:
[README] Add conda icon theme dependency

mpsonntag committed:
[setup] Update LICENCE file install path

mpsonntag committed:
[odml-gui] Reset used interpreter

mpsonntag committed:
[setup] Autogenerate ui startup script

Conda environments and virtual environments can crosspollute
the shebang line of the "odml-gui" startup script rendering
them unusable. Therefore use an autogenerated startup script instead.

mpsonntag committed:
[Helpers] Add conda environment check

mpsonntag committed:
[Editor] Refactor license file handling

mpsonntag committed:
[Editor] Add package and conda license paths

mpsonntag committed:
[Editor] Add get_img_path function

mpsonntag committed:
[Editor] Use get_img_path to load icons

mpsonntag committed:
[Editor] Add load icon existing dir check

mpsonntag committed:
[Editor] Fix loading odml-logo

mpsonntag committed:
[Editor] Remove old load image function

mpsonntag committed:
[Editor] Add conda environment to IconTheme paths

mpsonntag committed:
[Editor] Add virtual environment paths

achilleas-k commented on odmlui/Editor.py:
```diff
@@ -484,6 +536,78 @@ def map_tab(self, tab):
+            tail = os.path.split(uri_to_path(self.current_tab.file_uri))[1]
+            tmp_file = os.path.join(CACHE_DIR, ("%s.odml" % tail))
+            odml.fileio.save(self.current_tab.document, tmp_file)
+            os.system("odmltables -w compare -f %s &" % tmp_file)
```
Will this work if the filename has spaces in it?

achilleas-k commented on odmlui/Editor.py:
```diff
@@ -484,6 +536,78 @@ def map_tab(self, tab):
+            tail = os.path.split(uri_to_path(self.current_tab.file_uri))[1]
+            tmp_file = os.path.join(CACHE_DIR, ("%s.odml" % tail))
+            odml.fileio.save(self.current_tab.document, tmp_file)
+            os.system("odmltables -w convert -f %s &" % tmp_file)
```
Same as above. If the filename has spaces it will probably fail.

achilleas-k commented on odmlui/Editor.py:
```diff
@@ -484,6 +536,78 @@ def map_tab(self, tab):
+            tail = os.path.split(uri_to_path(self.current_tab.file_uri))[1]
+            tmp_file = os.path.join(CACHE_DIR, ("%s.odml" % tail))
+            odml.fileio.save(self.current_tab.document, tmp_file)
+            os.system("odmltables -w merge &")
```
Does the merge function not take file arguments?

JuliaSprenger replied:
Yes, you can provide one or two file names using -f (--file).

mpsonntag replied:
Then we shall provide the currently open document as the first file!

achilleas-k reviewed (commented):
Looks good, but maybe handle the filename issue for passing arguments to odmltables.

mpsonntag committed:
[Editor] Open doc as first odmltables merge file

mpsonntag committed:
[Helpers] Add run_odmltables function

mpsonntag committed:
[Editor] Use run_odmltables

mpsonntag committed:
[Editor] Reduce redundant odmltables code

mpsonntag committed:
[Helpers] Change odmltables system call

Use subprocess.Popen instead of os.system
in call to odmltables to conveniently handle
filenames containing whitespaces.

mpsonntag committed:
[Wizard] Fix missing term attribute

When creating an empty odml document, the
attribute self.section_page.term is not set.
Added a check to avoid page crash.
